### ISSUE TITLE
Unable to get deep linkin working, Linking.getInitialURL()

### ISSUE THREAD

**Issue by hannta** [wix/react-native-navigation#2017]
### Issue Description

I try to get deep linking working, user receives email with link. Url in that link is registered to my app. My app should start and I should get the link url. Sadly this is not working at all if my app is already running (in background).

_Btw. Official RNN [documentatio ](https://wix.github.io/react-native-navigation/#/deep-links) guides to use `onNavigatorEvent` event to receive deep links but I was not able to get that working at all?_

**App in background:**
- When I click link in email (gmail app), my app is launched/resumed, I see white screen and then it is minimized/moved to background. 
- No crashes or anything, app just disappears to background
- When I switch manually back to my app, I will get my deep link url using `Linking.getInitialURL()` as expected.

### Steps to Reproduce / Code Snippets / Screenshots

My AndroidMaifest.xml
```
<intent-filter >
	<action android:name="android.intent.action.VIEW" />
	<category android:name="android.intent.category.DEFAULT" />
	<category android:name="android.intent.category.BROWSABLE" />
	<data android:scheme="https"
		android:host="link.my.domain.com" />
</intent-filter>
```

My component which receives deep link events, im using AppState change events as described in #384
```
class MyComponent extends Component {
  constructor(props) {
    super(props);
    this.state = {
      currentAppState: AppState.currentState,
    };
  }

  componentDidMount() {
    AppState.addEventListener('change', this.handleAppStateChange);
  }

  componentWillUnmount() {
    AppState.removeEventListener('change', this.handleAppStateChange);
  }

  handleAppStateChange = (currentAppState) => {
    this.setState({ currentAppState });
    Linking.getInitialURL().then(url => this.handleDeepLink({ url }));
  };

  handleDeepLink = (event) => {
	  console.log(`Got event, url: ${event.url}`);

  };
  
  ...

```

---
### Environment
* React Native Navigation version: 1.1.235
* React Native version: 0.48.4
* Platform(s) (iOS, Android, or both?): Android
* Device info (Simulator/Device? OS version? Debug/Release?): Device OnePlus 5 & Nexus 5, debug

**Comment by srameshr** [comment]
Exactly, even I am facing the same issue. The listener for ```handleDeepLink``` never fires.

**Comment by gvillenave** [comment]
I just started experiencing a similar issue, in my case Linking.getInitialURL() always returns null whether the app is closed or in the background. It started happening after upgrading from RN 0.46.4 to 0.48.4 (and also to the last release of RNN). I wonder if it might be a React Native bug, perhaps https://github.com/facebook/react-native/issues/15992?

**Comment by joonhocho** [comment]
Are you guys using facebook sdk by any chance? https://github.com/facebook/react-native/pull/11674#issuecomment-290339775

**Comment by srameshr** [comment]
I use ```react-native-facebook-account-kit```

**Comment by hannta** [comment]
Im not using `react-native-facebook-account-kit` or any other linking libs.

In my case the issue is with launching/resuming app when I click the deep link url and the app is already running, app is not launched correctly. If I switch back to app manually, then `Linking.getInitialURL` is called correctly with my url (which I previously clicked).

**Comment by hannta** [comment]
Looks like I get this working if I set my MainActivity `android:launchMode="singleTask"`.

I was in impression that you should not use `singleTask` launch mode with RNN (ref: #1766 #384)?

_Updated my findings also to issue itself._

**Comment by ywongweb** [comment]
Add some logs in your `componentWillUnmount` and see if it's being called unexpectedly.

**Comment by gvillenave** [comment]
@joonhocho I'm using both `react-native-fbsdk` and `react-native-account-kit`. It doesn't seem like the PR you mentioned is getting any traction on Facebook's side unfortunately... Were you able to find any workaround?

**Comment by LittleLiByte** [comment]
i found this problem too.Linking.getInitialURL() always return null in release mode,but it's ok in debug mode.

**Comment by thatjuan** [comment]
Same issue on an ejected app. Not using any other module that handles urls (no facebook, no twitter, etc).

**Comment by perrosnk** [comment]
Any updates on this?

**Comment by bhatti-waqas** [comment]
@hannta Even if I use SingleTask as launch mode on android. I don't get the url in the app. If it was in background.

**Comment by brainahdev** [comment]
yes for me too ..I have used the FBSDK and even used SingleTask. Still not getting it working. please let us know

**Comment by pallavbakshi** [comment]
There has been no conclusive answer to this question of any opened/closed issue. Can we do something to escalate the request to look into the matter?

**Comment by theSMguy** [comment]
I'm having the same Issue, Expo detached Android app and SingleTask didn't solve the problem, any advances?

**Comment by joaoroque** [comment]
Using singleTask and placing this code on the splash screen solved this issue:

`componentDidMount () {
    Linking.addEventListener('url', this._handleOpenURL)
  }
  componentWillUnmount () {
    Linking.removeEventListener('url', this._handleOpenURL)
  }
  _handleOpenURL (event) {
    console.tron.log('SplashScreen _handleOpenURL: ' + event.url, true);
  }`

**Comment by perrosnk** [comment]
@joaoroque It might be a bit irrelevant to the question, but what do you mean by splash screen? Splash screen isn't it just an image? If not, could you please guide me on how to construct such a splash screen by sharing some code or give me a link to a source? Thanks!

**Comment by joaoroque** [comment]
@perrosnk The splash screen is our first screen. So you should use that code on the first screen that opens on your app.

**Comment by isaachinman** [comment]
Any update on this? I'm facing the same issue, except on iOS. `Linking.getInitialURL()` always returns `null`, despite external URIs actually opening my app on simulator.

**Comment by jdanieldev** [comment]
@joaoroque your solution works for the url problem, but when _handleOpenURL is executed, this = undefined, so you can't use nothing on the class. Very weird case.

**Comment by isaachinman** [comment]
Complete failure on the maintainer's part to even _respond_ to this issue after almost 8 months, and now it's marked "stale".

With over 700 open issues, it's really a shame that this lib is still the best option for RN navigation.

It's not a matter of motivated users - there are also over 60 open PRs and I assume many others in this particular thread, including myself, would be happy to fix and submit a PR for this issue if its root cause was clear.

**Comment by michaelsyap** [comment]
Hi everyone, 

I think I solved this issue. Here's how I did it.

I used `Linking` and `AppState` from `react-native` in the root class of the app that runs first before you trigger your root screen through `Navigation.startSingleScreenApp`. In my case it's called `App.js`.

Note 1: I followed instructions to add `Linking` support to your app [**here**](https://facebook.github.io/react-native/docs/linking.html#constructor).

Note 2: Focus on how I implemented it by checking these following functions:

- `bootstrapDeepLinks`
- `_handleOpenURL`
- `_handleAppStateChange`
- `_checkInitialURL`

Note 3: This code is made to demonstrate capturing deep links from the browser (through `console.warn`) in android and IOS with the following use cases below:

- Capture deeplink upon opening the app that is closed from the web browser 
- Capture deeplink upon returning to the app while in the background  from the web browser
- Make sure that deeplink updates doesn't fire again when navigating to another screen after the app has been triggered open (either from closed state or from background state) from a deeplink.

`App.js`:
```js
import React from 'react';
import { Provider } from 'react-redux';
import { persistStore, persistReducer } from 'redux-persist';


import store from 'src/stores';
import { triggerRootScreen } from 'src/actions/navigation';

import { Linking, AppState, Platform } from 'react-native';

import { registerScreens } from './navigation/screens';
import { Navigation } from 'react-native-navigation';
import { iconsMap, iconsLoaded } from 'src/config/icons';

import { BRAND_PRIMARY_LIGHT, BRAND_GRAY_LIGHT, BRAND_ACCENT } from 'src/config/colors';


const iconInsets =  { // add this to change icon position (optional, iOS only).
  top: 6, // optional, default is 0.
  left: 0, // optional, default is 0.
  bottom: -6, // optional, default is 0.
  right: 0 // optional, default is 0.
};

const navigatorStyle =  {
  navBarHidden: true, 
  drawUnderStatusBar: true,
}

class App extends React.Component {
  _appState = '';

  constructor(props) {
    super(props);

    this.initBootstrapApp();
  }
  

  onStoreUpdate() {
    const reduxState = store.getState();
    console.log(reduxState.navigation.currentRootScreen);
    console.log(reduxState.userAuth);
    
    // If the current root navigator being loaded is not the same as the loaded root navigator
    if(this.currentRootScreen !== reduxState.navigation.currentRootScreen) {
      
      this.currentRootScreen = reduxState.navigation.currentRootScreen;

      
      if(reduxState.userAuth.user_status === 'SIGNED_IN' ) {
        iconsLoaded.then(() => {
          this.initMainApp();
        });
        return;
      }

      switch(reduxState.navigation.currentRootScreen) {
        case 'ONBOARDING':
          this.initOnboarding();
          return;
        case 'USER_GATE':
          this.initUserGate();
          return;
        case 'SIGN_IN':
          this.initSigninGate();
          return;
        case 'REGISTRATION':
          this.initRegistrationGate();
          return;
        case 'MAIN_APP':
          iconsLoaded.then(() => {
            this.initMainApp();
          });
          return;
        default:
          this.initUserGate();
      }

    }

    return;
  }


  initBootstrapApp() {
    this.bootstrapDeepLinks();
    this.bootstrapPersistStore();
  }

  bootstrapPersistStore() {
    persistStore(store, null, (() => {
      console.log('Redux Persist storage done loading...');

      this.currentRootScreen = '';
      store.subscribe(this.onStoreUpdate.bind(this));

      registerScreens(store, Provider);

      this.onStoreUpdate();
      // store.dispatch(triggerRootScreen('SIGN_IN'));
    }).bind(this));
  }

  bootstrapDeepLinks() {

    if(Platform.OS === 'android') {
      //This will watch for updates in the deeplink in android. This will also fire on first app load.
      AppState.addEventListener('change', this._handleAppStateChange.bind(this));
    } else {
      //This will watch for updates in the deeplink in ios. This will not fire on first app load.
      Linking.addEventListener('url', this._handleOpenURL);
      
      //Fire on first load in ios
      this._checkInitialURL()
    }
  }


  // This will check for deeplink changes in ios only
  _handleOpenURL(event) {
    
    if(__DEV__) {
      console.warn('Updated url in IOS is: ' + event.url);
    }

  }

  // This will check for deeplink changes in android only
  _handleAppStateChange(currentAppState) {
    this._appState = currentAppState;

    this._checkInitialURL();
  }

  _checkInitialURL() {
    
    Linking.getInitialURL().then((url) => {
      
      if (url) {
      
        if(__DEV__) {

          if(Platform.OS === 'android') {
            console.warn('Updated url in Android is: ' + url);
          } else {
            // Linking.getInitialURL() will only be fired once in IOS
            console.warn('Initial url is: ' + url);
          }
        }

      }
    }).catch(err => console.warn('An error occurred', err));

  }


  initMainApp() {


    Navigation.startTabBasedApp({
			tabs: [
				{
					// label: 'Home',
					screen: 'sampleApp.MyAccountPage',
					icon: iconsMap['ico_profile'],
          selectedIcon: iconsMap['ico_profile'],
          iconInsets,
          navigatorStyle
				},
				{
					// label: 'Search',
					title: 'Menu',
					screen: 'sampleApp.SearchPage',
					icon: iconsMap['ico_glass'],
					selectedIcon: iconsMap['ico_glass'],
          iconInsets,
          navigatorStyle
				},
				{
					// label: 'Home',
					// title: 'Home',
					screen: 'sampleApp.HomePage',
					icon: iconsMap['ico_home'],
					selectedIcon: iconsMap['ico_home'],
          iconInsets,
          navigatorStyle
				},
				{
					// label: 'Cart',
					title: 'Cart',
					screen: 'sampleApp.CartPage',
					icon: iconsMap['ico_cart'],
					selectedIcon: iconsMap['ico_cart'],
          iconInsets,
          navigatorStyle
				},
				{
					// label: 'Menu',
					title: 'Menu',
					screen: 'sampleApp.ProductsCategories',
					icon: iconsMap['ico_menu'],
					selectedIcon: iconsMap['ico_menu'],
          iconInsets,
          navigatorStyle
				}
      ],
      tabsStyle: {
        initialTabIndex: 0, // optional, the default selected bottom tab. Default: 0. On Android, add this to appStyle
        tabBarButtonColor: BRAND_GRAY_LIGHT,
        tabBarSelectedButtonColor: BRAND_PRIMARY_LIGHT
      },
      appStyle: {
        initialTabIndex: 0,
        bottomTabBadgeBackgroundColor: BRAND_ACCENT, // Optional, change badge background color. Android only
        tabBarButtonColor: BRAND_GRAY_LIGHT, // For Android Config
        tabBarSelectedButtonColor: BRAND_PRIMARY_LIGHT //For Android Config
      }
    });

  }

  initOnboarding() {
    Navigation.startSingleScreenApp({
      screen: {
        screen: 'sampleApp.Onboarding',
        navigatorStyle: {
          navBarHidden: true,
          drawUnderStatusBar: true
        }
      }
    })
  }

  initUserGate() {
    Navigation.startSingleScreenApp({
      screen: {
        screen: 'sampleApp.UserGate',
        navigatorStyle: {
          navBarHidden: true,
          drawUnderStatusBar: true
        }
      }
    });

  }


  initSigninGate() {

    Navigation.startSingleScreenApp({
      screen: {
        screen: 'sampleApp.SigninGate',
        navigatorStyle: {
          navBarHidden: true,
          drawUnderStatusBar: true
        }
      }
    });

  }

  initRegistrationGate() {

    Navigation.startSingleScreenApp({
      screen: {
        screen: 'sampleApp.RegistrationGate',
        navigatorStyle: {
          navBarHidden: true, 
          drawUnderStatusBar: true,
          drawUnderNavBar: true,  
          navBarTransparent: true
        }
      }
    });
  }

  

}


export default App;
```

I hope this helps!

**Comment by witalobenicio** [comment]
@michaelsyap on iOS how you define what URLs to listen?

**Comment by fabiancarlos** [comment]
Hi guys, wrap yours deeplink handle on a `InteractionManager` like so:
`
componentDidMount() {
  InteractionManager.runAfterInteractions(() => {
     this._addLinkingListener();
  });
}
`
Dont forget to import `InteractionManager` on react-native.

**Comment by michaelsyap** [comment]
@witalobenicio I defined a deep link to something like this  sampleapp://action?screen=payment

For more information I just followed instructions here https://medium.com/react-native-training/deep-linking-your-react-native-app-d87c39a1ad5e

**Comment by witalobenicio** [comment]
Hi @michaelsyap and @fabiancarlos I was able to make it work with RNN using just React Native Linking, but after that, my team end up deciding to use Branch Links as solution because we need other things beside the deeplinking. Thank you.

**Comment by joelprakash** [comment]
Hi Guys,

I've found a solution for React Native. This solution is for an app which is already running, to get deep links to trigger some logic.

Issue:
This issue occurs, when the app is already opened and running. I am sure many of you have defined your deeplink handler such as 'Linking.getInitialURL()' for Android or 'Linking.addEventListener('url', this.handleOpenURL)' on ios side inside 'componentDidMount()'. Now since it is inside componentDidMount it will only trigger if an event occurs. One example of this are button clicks that trigger events and so forth. As you may have noticed certain navigations and putting the app on background or bringing the screen to focus does not trigger the componentDidMount function. Currently there are no in-built screen lifecycle methods that deal with screen focus or active screen and so on.
Due to this your app may open when you deeplink due to the efforts by android and ios OS. But other than that, you will see that no logic or event will occur. The app seems not to respond to the deeplink after opening.

Solution:
We will remove all deeplink event handlers from componentDidMount() and instead use AppState from react-native default library in componentDidMount. AppState will change the state of the app each time the screen becomes infocus or active. This in turn will call the deeplink event handlers and allow them to trigger.

import { Linking, AppState } from 'react-native';

constructor(props) {
        super(props);
        this.state = {currentAppState: AppState.currentState};
        this._handleAppStateChange = this._handleAppStateChange.bind(this);
}

 componentDidMount() {
	AppState.addEventListener('change', this._handleAppStateChange);
	if (Platform.OS === 'android') {this._deepLinkForAndroid();
	}else {this._deepLinkForIos();}
}

componentWillUnmount() {
      AppState.removeEventListener('change', this._handleAppStateChange);
}

_handleAppStateChange(currentAppState) {
     this.setState({currentAppState});
     if (Platform.OS === 'android') {this._deepLinkForAndroid();
     }else {this._deepLinkForIos();}
}

_deepLinkForAndroid() {
    Linking.getInitialURL().then(url => {this.navigate(url); });
}
	
_deepLinkForIos() {
    Linking.addEventListener('url', this.handleOpenURL);
}

Also on Android make sure you don't have an 'android:launchMode="singleTask"' defined inside your MainActivity

For ios make sure you add the lines in your 'AppDelegate.m' for listening for incoming deeplinks when you app is already running as seen here https://facebook.github.io/react-native/docs/linking

I'm no expert on react. So, feel free to correct or add to what I have. I hope my solution above is able to help you out.

**Comment by murat-mehmet** [comment]
**Found solution!**
I have been struggling with this for a day and since there is no support from officials, I had to walk in source code for hours. Finally I found out that there's no way to make getInitialURL() work in v1 (or getInitialLink() in react-native-firebase). Because the initial intent is emitted as an event and gets cleared immediately on activity onResume method. 
I was adding the listener in componentWillMount() like most people, which was too late to listen for that event, because onResume is called much before js methods.

Long story short, you must add listener in App.js, before starting any component, which then it is called as soon as app starts. You can save it to a global static variable and emit as event to whole app, then check the global variable in componentWillMount to find initial link..

Or use this simple helper class:
```
import firebase from "react-native-firebase";
import EventEmitter from 'EventEmitter';

export default class LinkingHelper {
    static listenLinks() {
        firebase.links().onLink(link => {
            if (waitingInitialLink)
                initialLink = link;
            emitter.emit('url', link);
            console.log('appjs link', link);
        });
    }

    static getInitialLink() {
        return initialLink;
    }

    static onLink(listener) {
        waitingInitialLink = false;
        emitter.addListener('url', listener, null);
        return () => emitter.removeListener('url', listener);
    }
}

const emitter = new EventEmitter(null);
let initialLink = null;
let waitingInitialLink = true;
```
**Usage:**
Call `LinkingHelper.listenLinks();` before `Navigation.startSingleScreenApp`
Later in your component: 
```
    componentWillMount() {
        const initialLink = LinkingHelper.getInitialLink();
        if (initialLink){
            console.log('Initial link: ', initialLink);
        }
        this.linkListener = LinkingHelper.onLink(link => {
            console.log('On link', link);
        });
    }
    componentWillUnmount() {
            this.linkListener();
    }
```

Note: To work with RN Linking, you need to change `firebase.links().onLink(...)` with `Linking.addEventListener('url',...)`. I only tested with firebase.
